drivers/raw_exec/driver.go

```diff
@@ -0,0 +1,530 @@
+	// that will immediately be sent
+	fingerprintCh chan *drivers.Fingerprint
+
+	stopCh chan struct{}
+
+	logger hclog.Logger
+}
+
+type Config struct {
+	// NoCgroups tracks whether we should use a cgroup to manage the process
+	// tree
+	NoCgroups bool `codec:"no_cgroups"`
+
+	// Enabled is set to true to enable the raw_exec driver
+	Enabled bool `codec:"enabled"`
+}
+
+type TaskConfig struct {
+	Command string   `codec:"command"`
+	Args    []string `codec:"args"`
+}
+
+func NewRawExecDriver(logger hclog.Logger) drivers.DriverPlugin {
+	stopCh := make(chan struct{})
+	return &RawExecDriver{
+		Eventer:       utils.NewEventer(stopCh),
+		config:        &Config{},
+		tasks:         newTaskStore(),
+		fingerprintCh: make(chan *drivers.Fingerprint),
+		stopCh:        stopCh,
+		logger:        logger.Named(pluginName),
+	}
+}
+
+func (r *RawExecDriver) PluginInfo() (*base.PluginInfoResponse, error) {
+	return pluginInfo, nil
+}
+
+func (r *RawExecDriver) ConfigSchema() (*hclspec.Spec, error) {
+	return configSpec, nil
+}
+
+func (r *RawExecDriver) SetConfig(data []byte) error {
+	var config Config
+	if err := base.MsgPackDecode(data, &config); err != nil {
+		return err
+	}
+
+	r.config = &config
+	go r.fingerprintNow()
+	return nil
+}
+
+func (r *RawExecDriver) Shutdown(ctx context.Context) error {
+	close(r.stopCh)
+	return nil
+}
+
+func (r *RawExecDriver) TaskConfigSchema() (*hclspec.Spec, error) {
+	return taskConfigSpec, nil
+}
+
+func (r *RawExecDriver) Capabilities() (*drivers.Capabilities, error) {
+	return capabilities, nil
+}
+
+func (r *RawExecDriver) Fingerprint(ctx context.Context) (<-chan *drivers.Fingerprint, error) {
+	ch := make(chan *drivers.Fingerprint)
+	go r.fingerprint(ctx, ch)
+	return ch, nil
+}
+
+func (r *RawExecDriver) fingerprint(ctx context.Context, ch chan *drivers.Fingerprint) {
+	defer close(r.fingerprintCh)
+
+	ticker := time.NewTimer(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			ticker.Reset(fingerprintPeriod)
+			go r.fingerprintNow()
+		case f := <-r.fingerprintCh:
+			ch <- f
+		}
+	}
+}
+
+func (r *RawExecDriver) fingerprintNow() {
+	if r.fingerprintCh == nil {
+		r.logger.Debug("fingerprint channel was nil, skipping fingerprint")
+		return
+	}
+	var health drivers.HealthState
+	var desc string
+	attrs := map[string]string{}
+	if r.config.Enabled {
+		health = drivers.HealthStateHealthy
+		desc = "raw_exec enabled"
+		attrs["driver.raw_exec"] = "1"
+	} else {
+		health = drivers.HealthStateUndetected
+		desc = "raw_exec disabled"
+	}
+	r.fingerprintCh <- &drivers.Fingerprint{
+		Attributes:        map[string]string{},
+		Health:            health,
+		HealthDescription: desc,
+	}
+}
+
+func (r *RawExecDriver) RecoverTask(handle *drivers.TaskHandle) error {
+	var taskState RawExecTaskState
+
+	err := handle.GetDriverState(&taskState)
+	if err != nil {
+		r.logger.Error("failed to recover task", "error", err, "task_id", handle.Config.ID)
```

Comment:
All these log messages are the same. They should be more descriptive:
`failed to get driver state from handle`